### PR TITLE
Enhance jenkins-{persistent,ephemeral}-template

### PR DESCRIPTION
**Pull request by tnozicka**
This patch allows you to:
 - use custom ImageStream
 - have better control over setting limits
 - have multiple Jenkins instances in one namespace

### CODE REVIEW

File: examples/jenkins/jenkins-persistent-template.json [hunk 1]
```diff
@@ -204,11 +195,35 @@
       "value": "password"
     },
     {
+      "name": "MEMORY_REQUEST",
+      "displayName": "Memory Lower Limit",
+      "description": "Reserved amount of memory for the container.",
+      "value": "1200Mi"
+    },
```
Comment (bparees) [inline examples/jenkins/jenkins-persistent-template.json]: We don't allow users to set requests in online, why do you feel you need to do this?

File: examples/jenkins/jenkins-persistent-template.json [hunk 2]
```diff
@@ -204,11 +195,35 @@
...
+    {
+      "name": "MEMORY_LIMIT",
+      "displayName": "Memory Upper Limit",
+      "description": "Maximum amount of memory the container can use.",
+      "value": "2000Mi"
+    },
```
Comment (bparees) [inline examples/jenkins/jenkins-persistent-template.json]: this isn't an acceptable value for online, it would consume the user's entire memory quota.  why do you think the value needs to be this high by default?

File: examples/jenkins/jenkins-persistent-template.json [hunk 3]
```diff
@@ -204,11 +195,35 @@
...
+    {
+      "name": "JENKINS_IS_NAMESPACE",
+      "displayName": "Jenkins ImageStream Namespace",
+      "description": "The OpenShift Namespace where the Jenkins ImageStream resides.",
+      "value": "openshift"
+    },
+    {
+      "name": "JENKINS_IST",
+      "displayName": "Jenkins ImageStreamTag",
+      "description": "Tag (name) of the ImageStream in that namespace.",
```
Comment (bparees) [inline examples/jenkins/jenkins-persistent-template.json]: "Name of the ImageStreamTag to be used for the Jenkins image."

and the parameter name should be JENKINS_IMAGE_STREAM_TAG, users will be confused by "IST"

File: examples/jenkins/jenkins-persistent-template.json [hunk 4]
```diff
@@ -204,11 +195,35 @@
...
+    {
       "name": "VOLUME_CAPACITY",
       "displayName": "Volume Capacity",
-      "description": "Volume space available for data, e.g. 512Mi, 2Gi.",
+      "description": "Volume space available for data, e.g. 500Mi, 2Gi.",
       "value": "1Gi",
       "required": true
+    },
+    {
+      "name": "JENKINS_IS_NAMESPACE",
```
Comment (bparees) [inline examples/jenkins/jenkins-persistent-template.json]: We just call this NAMESPACE in our other templates, let's stay consistent:
https://github.com/openshift/origin/blob/master/examples/db-templates/mongodb-persistent-template.json#L81

**Comment by bparees** [comment]
these changes need to be applied to the ephemeral template as well.

File: examples/jenkins/jenkins-persistent-template.json (hunk 1 above)
Comment (tnozicka) [inline examples/jenkins/jenkins-persistent-template.json]: I would like to have a burstable instance, but if online can't do this I will remove it.

Comment (bparees) [inline examples/jenkins/jenkins-persistent-template.json]: request gets set implicitly, so it's still a burstable instance, effectively.

File: examples/jenkins/jenkins-persistent-template.json (hunk 3 above)
Comment (tnozicka) [inline examples/jenkins/jenkins-persistent-template.json]: ok, will rephrase

File: examples/jenkins/jenkins-persistent-template.json (hunk 2 above)
Comment (tnozicka) [inline examples/jenkins/jenkins-persistent-template.json]: This should be just the upper limit. This memory would not be allocated unless needed. But I did not know about online forbidding setting requests. In that case which value do you propose? I am inclined to a higher value because OpenShift web console does not tell you that you container has been killed by OOM killer due to this setting (AFAIK). I have seen 2 defaults: 500Mi and 1000Mi. As I have tried different Jenkins containers, e.g. the modified one from docker.io/library/jenkins it has been killed by OOM killer (500Mi - every time, 1Gi - ~2) while initializing.

File: examples/jenkins/jenkins-persistent-template.json (hunk 4 above)
Comment (tnozicka) [inline examples/jenkins/jenkins-persistent-template.json]: ok

File: examples/jenkins/jenkins-persistent-template.json (hunk 2 above)
Comment (bparees) [inline examples/jenkins/jenkins-persistent-template.json]: you are quota'd against your upper limit.  if you have a 2gig quota, you can only deploy resources with a total of 2 gigs of upper limit.  So if a user deploy this, they would not be able to deploy anything else in online.  

We need to keep this limit to 512meg.  It's been validated to work with our jenkins image, if you're trying to use another image, you shouldn't really be using this template anyway. This template defines env vars that are respected specifically by our image.

Would like to know why you're using the dockerhub jenkins image instead of our image?

Comment (bparees) [inline examples/jenkins/jenkins-persistent-template.json]: btw, this is why our image works when the container is resource limited and the docker hub image does not:
https://github.com/openshift/jenkins/blob/master/1/contrib/s2i/run#L16-L27

we set the java max heap appropriately based on the container memory constraints.

Comment (tnozicka) [inline examples/jenkins/jenkins-persistent-template.json]: > you are quota'd against your upper limit.
ok let's keep it at 500Mi

> Would like to know why you're using the dockerhub jenkins image instead of our image?

Our image has old version of Jenkins, that's the only reason. But my point with this template was just to get specification of different images streams working.

**Comment by tnozicka** [comment]
@bparees please take another look. I have fixed it according to your suggestions.
I will update the ephemeral template after you decide if these changes are fine.

**Comment by bparees** [comment]
looks good.

**Comment by bparees** [comment]
actually not quite looks good, put the memory default back to 512 please.

**Comment by tnozicka** [comment]
@bparees 512Mi seems really suspicious to me. **500Mi = 512M**. If you are worried about backwards compatibility I would suggest using some bigger number like 520Mi. Using 2^x with _Mi_ seem strange a leads you to think that the people actually meant to use _M_ instead of _Mi_.

e.g. kubernetes doc does not use the 2^x numbers with Mi http://kubernetes.io/docs/admin/limitrange/

**Comment by bparees** [comment]
paging @gabemontero to the thread as he was the one who worked on the team that came up with these defaults originally.

**Comment by gabemontero** [comment]
Well, after circling back to [the doc here](https://docs.openshift.org/latest/dev_guide/compute_resources.html#viewing-compute-resources) and surrounding sections, I'll admit I'm a bit uncertain now.

History:
1) in all the examples I've seen, `Mi` or `Gi` is used for memory wrt quotas; I had presumed those were the only suffixes allowed ... @derekwaynecarr - are we even allowed to use `G` or `M` when specifying amounts of memory wrt quota

2) after a bunch of fine tuning of these numbers both with the jenkins template as well as the others for openshift online, @danmcp ultimately directed me to make them universally `512` to line up with default memory allocation on AWS (and I believe he hit the merge button for the PRs which did this); assuming that only `Mi` was allowed, I left the suffix the same.

Ultimately, whether it is 512 or 500 or M or Mi, we want the default to line up with what the underlying IaaS.

**Comment by danmcp** [comment]
@gabemontero https://github.com/openshift/origin/pull/7707

I don't think the IaaS is the important factor here as much as the project quota.  n containers at the default size should fit exactly in the quota.

**Comment by gabemontero** [comment]
thanks for the clarification @danmcp

So based on https://github.com/openshift/online/blob/master/templates/project-request.json the project quota for memory is 2 Gi.

Then, looking at https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/resource/quantity.go, it does look like you can specify M vs. Mi.

However, looking at https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/resource/quantity.go#L69, if 1.5 Gi translates to 1536 Mi. that is 512Mi *3, so 512Mi * 4 is 2 Gi.

(note, the compute-resources-timebound hard limit for online is 1536Mi).

Aside from the comment I noted, looking at the BinarySi (i.e. Mi or Gi) and DecimalSi, that seems to still line up with the comment.

Bottom line - I believe it needs to stay at 512Mi

**Comment by tnozicka** [comment]
@gabemontero @danmcp Thanks for the clarification. I have put it back to 512Mi.

@bparees I have have fixed the ephemeral template as well, please review

**Comment by bparees** [comment]
lgtm, please squash the commits and i'll merge.

**Comment by tnozicka** [comment]
@bparees squashed

**Comment by bparees** [comment]
thanks @tnozicka 
[merge]
